### PR TITLE
Add support for pagination

### PR DESCRIPTION
**Pull request by aashikam**
# Description

fixes: https://github.com/ballerina-platform/ballerina-library/issues/6577
fixes: https://github.com/ballerina-platform/ballerina-library/issues/6548

- [x] Test for maxRecords less than the number of records
- [x] Test for maxRecords more than the number of records
- [x] Add a changelog
- [x] Doc changes

### CODE REVIEW

File: ballerina/client.bal [hunk 1]
```diff
@@ -651,18 +652,55 @@ public isolated client class Client {
...
+    # + maxRecords - The maximum number of records to retrieve per set of results for the query
+    # + return - The resulting string[][] if successful else `error`
+    isolated remote function getQueryResult(string bulkJobId, int? maxRecords = ())
             returns string[][]|error {
-        string path = utils:prepareUrl([API_BASE_PATH, JOBS, QUERY, bulkJobId, RESULT]);
+                
+        string path = "";
+        string batchingParams = "";
+
+        // Max records value is user given
```
Comment (NipunaRanasinghe) [inline ballerina/client.bal:666]: IMO this comment is redundant
```suggestion
```

File: ballerina/client.bal [hunk 2]
```diff
@@ -651,18 +652,55 @@ public isolated client class Client {
...
 
-    # Get query job results
+    # Get bulk query job results
+    # 
     # + bulkJobId - Id of the bulk job
-    # + return - string[][] if successful else `error`
-    isolated remote function getQueryResult(string bulkJobId)
+    # + maxRecords - The maximum number of records to retrieve per set of results for the query
+    # + return - The resulting string[][] if successful else `error`
+    isolated remote function getQueryResult(string bulkJobId, int? maxRecords = ())
```
Comment (NipunaRanasinghe) [inline ballerina/client.bal:660]: We better make this a defaultable parameter instead (default value should be equal to the salesforce default)
```suggestion
    isolated remote function getQueryResult(string bulkJobId, int maxRecords = 50000)
```

File: ballerina/client.bal [hunk 3]
```diff
@@ -651,18 +652,55 @@ public isolated client class Client {
...
+                
+        string path = "";
+        string batchingParams = "";
+
+        // Max records value is user given
+        if maxRecords != () {
+            lock {
+                if self.sfLocators.hasKey(bulkJobId) {
+                    string locator = self.sfLocators.get(bulkJobId);
+                    if locator is "null" {
```
Comment (NipunaRanasinghe) [inline ballerina/client.bal:671]: How can the locator be a "null" string? is that what we get from the header?

File: ballerina/client.bal [hunk 4]
```diff
@@ -651,18 +652,55 @@ public isolated client class Client {
...
+        } 
+        
         http:Response response = check self.salesforceClient->get(path);
         if response.statusCode == 200 {
             string textPayload = check response.getTextPayload();
             if textPayload == "" {
                 return [];
             }
+            lock {
+                self.sfLocators[bulkJobId] = check response.getHeader("sforce-locator");
```
Comment (NipunaRanasinghe) [inline ballerina/client.bal:702]: Shouldn't we gracefully handle the error in this case? and, cant we assume that the locator is null if an error occurs?

**Review by NipunaRanasinghe — CHANGES_REQUESTED** [review]

File: ballerina/client.bal (hunk 3 above)
Comment (aashikam) [reply ballerina/client.bal:671]: yup

File: ballerina/client.bal (hunk 4 above)
Comment (aashikam) [reply ballerina/client.bal:702]: it returns "null"

File: ballerina/client.bal [hunk 5]
```diff
@@ -651,18 +652,54 @@ public isolated client class Client {
...
+                        return [];
+                    } 
+                    batchingParams = string `results?maxRecords=${maxRecords}&locator=${locator}`;
+                } else {
+                    batchingParams = string `results?maxRecords=${maxRecords}`;
+                }
+            }
+            path = utils:prepareUrl([API_BASE_PATH, JOBS, QUERY, bulkJobId, batchingParams]);
+            // Max records value default, we might not know when the locator comes
+        } else if maxRecords == () {
```
Comment (NipunaRanasinghe) [inline ballerina/client.bal:680]:
```suggestion
        } else {
```

File: ballerina/client.bal [hunk 6]
```diff
@@ -651,18 +652,54 @@ public isolated client class Client {
...
+            path = utils:prepareUrl([API_BASE_PATH, JOBS, QUERY, bulkJobId, batchingParams]);
+            // Max records value default, we might not know when the locator comes
+        } else {
+            lock {
+                if self.sfLocators.hasKey(bulkJobId) {
+                    string locator = self.sfLocators.get(bulkJobId);
+                    if locator is "null" {
+                        return [];
+                    } 
+                    batchingParams = string `results?locator=${locator}`;
```
Comment (NipunaRanasinghe) [inline ballerina/client.bal:687]: it seems we are not setting the `path` within this if statement (hence no request to be made). Is that expected?

Comment (aashikam) [reply ballerina/client.bal:687]: Nope, my mistake. 😃 Fixed it

**Review by NipunaRanasinghe — APPROVED** [review]
LGTM

**Pull request merged**